File: src-python/src/services/cache.py
"""In-memory cache service with TTL support."""

import threading
import time
from dataclasses import dataclass
from typing import Any


@dataclass
class CacheEntry:
    """A cached value with expiration time."""

    value: Any
    expires_at: float | None  # None means no expiration


class CacheService:
    """Thread-safe in-memory cache with TTL support."""

    def __init__(self) -> None:
        """Initialize the cache service."""
        self._cache: dict[str, CacheEntry] = {}
        self._lock = threading.RLock()

    def get(self, key: str) -> Any | None:
        """Get a value from the cache.

        Args:
            key: The cache key

        Returns:
            The cached value, or None if not found or expired
        """
        with self._lock:
            entry = self._cache.get(key)
            if entry is None:
                return None

            # Check expiration
            if entry.expires_at is not None and time.time() > entry.expires_at:
                del self._cache[key]
                return None

            return entry.value

    def set(self, key: str, value: Any, ttl: float | None = None) -> None:
        """Set a value in the cache.

        Args:
            key: The cache key
            value: The value to cache
            ttl: Time-to-live in seconds (None for no expiration)
        """
        with self._lock:
            expires_at = None if ttl is None else time.time() + ttl
            self._cache[key] = CacheEntry(value=value, expires_at=expires_at)

    def delete(self, key: str) -> bool:
        """Delete a value from the cache.

        Args:
            key: The cache key

        Returns:
            True if the key was deleted, False if not found
        """
        with self._lock:
            if key in self._cache:
                del self._cache[key]
                return True
            return False

    def clear(self) -> None:
        """Clear all values from the cache."""
        with self._lock:
            self._cache.clear()

    def cleanup_expired(self) -> int:
        """Remove all expired entries from the cache.

        Returns:
            Number of entries removed
        """
        with self._lock:
            now = time.time()
            expired_keys = [
                key
                for key, entry in self._cache.items()
                if entry.expires_at is not None and now > entry.expires_at
            ]
            for key in expired_keys:
                del self._cache[key]
            return len(expired_keys)

File: src-go/internal/cache/cache.go
// Package cache provides a thread-safe in-memory cache with TTL support.
package cache

import (
	"sync"
	"time"
)

// Entry represents a cached value with optional expiration.
type Entry struct {
	Value     interface{}
	ExpiresAt *time.Time // nil means no expiration
}

// Cache is a thread-safe in-memory cache with TTL support.
type Cache struct {
	mu    sync.RWMutex
	items map[string]Entry
}

// New creates a new cache instance.
func New() *Cache {
	return &Cache{
		items: make(map[string]Entry),
	}
}

// Get retrieves a value from the cache.
// Returns nil if not found or expired.
func (c *Cache) Get(key string) interface{} {
	c.mu.RLock()
	entry, ok := c.items[key]
	c.mu.RUnlock()

	if !ok {
		return nil
	}

	// Check expiration
	if entry.ExpiresAt != nil && time.Now().After(*entry.ExpiresAt) {
		c.Delete(key)
		return nil
	}

	return entry.Value
}

// Set stores a value in the cache with an optional TTL.
// If ttl is 0 or negative, the value never expires.
func (c *Cache) Set(key string, value interface{}, ttl time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()

	entry := Entry{Value: value}
	if ttl > 0 {
		expiresAt := time.Now().Add(ttl)
		entry.ExpiresAt = &expiresAt
	}

	c.items[key] = entry
}

// Delete removes a value from the cache.
// Returns true if the key was deleted, false if not found.
func (c *Cache) Delete(key string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()

	if _, ok := c.items[key]; ok {
		delete(c.items, key)
		return true
	}
	return false
}

// Clear removes all values from the cache.
func (c *Cache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.items = make(map[string]Entry)
}

// CleanupExpired removes all expired entries from the cache.
// Returns the number of entries removed.
func (c *Cache) CleanupExpired() int {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := time.Now()
	count := 0

	for key, entry := range c.items {
		if entry.ExpiresAt != nil && now.After(*entry.ExpiresAt) {
			delete(c.items, key)
			count++
		}
	}

	return count
}
